### DISCUSSION TITLE
Secure Connection Failed

### DISCUSSION THREAD

**Discussion by Shadow8472** [dani-garcia/vaultwarden#2149] [Q&A] [answered]
I am so close, but it feels like this project is fighting me to the bitter end.

- podman run -d --name vaultwarden --restart unless-stopped
- -v /home/vaultwardenUsr/<path/to/vw-data>/:/data/:Z 
- -v /home/vaultwardenUsr/<path/to/private/certs>/:/ssl/:Z
- -e ROCKET_TLS='{certs="/ssl/bitwarden.crt",key="/ssl/bitwarden.key"}' 
- -p 44300:443
- vaultwarden/server:latest

I'm trying to rule out points of failure
- Server's firewall has port 44300 open for TCP, but not UDP.
- /home/vaultwardenUsr/<path/to/private/certs>/ only contains bitwarden.crt and bitwarden.key
- Vaultwarden container can run for hours (without the bitwarden .crt and .key files, podman ps always lists it as started "less than a second ago,").
- I've imported my certificate to Firefox.
- I'm trying to go to https://10.0.0.X:44300
The most useful error Firefox is giving is PR_CONNECT_RESET_ERROR 

I previously set up Bitwarden_RS and that server suffered corruption. I've been living off Firefox Bitwarden plugins since then. While testing, I exported my passwords to migrate and logged out on one of my workstations.

Do I need any existing certificates in /etc/ssl/certs/? (I doubt it. As I worked, I found those are only symbolic links; when the container loads symbolic links, they point to the container's internal structure.)

I've looked up what error codes I'm getting, and I can't tell where the problem is. My only guess is that the container should be forwarding traffic to port 80 only, even with HTTPS, but that doesn't feel right.

**Comment by Masgalor** [comment, marked as answer]
While I have no clue about docker, or podman, I may give you some generic advice.

The error you mentioned ´PR_CONNECT_RESET_ERROR´ is so broad that it is almost uselss.
Its message is, that the connection was forcefully reset while establishing a secure channel.
Some usual reasons for this are:
 - Server and client could not negotiate a common protocol
 - Server and client could not negotiate a common cipher
 - Server or client cant build a trust chain because something is missing
 - Some firewall is intercepting the connection

The best advice I can give you, is the same as the wiki and @BlackDex already gave you.
_**Do not use ROCKET for TLS.**_

Use an established reversproxy as nginx or caddy, make them serve vaultwarden on plain http and after you are sure the proxy works correct add https to your existing configuration.

Anyway, we need to see the logs of ROCKET to help you any further.

**Reply by Shadow8472** [reply]
Perhaps I will look into using a reverse proxy after all. I went ahead and changed the container's forwarded port from 443 to 80 and it worked. I'm essentially back where I was before my Pi's SD corrupted except I'm still down a plugin. I suppose I can go ahead and restore that functionality and press on. A reverse proxy will do future projects well too I suppose.

My question earlier as I understand it was essentially "Why not 443?" All the tutorials I've seen forward to port 80, but none explain why not 443 when setting up ROCKET as I was. (I was semi-following one by sensiCLICK on YouTube.)

**Reply by Masgalor** [reply]
> All the tutorials I've seen forward to port 80, but none explain why not 443

ROCKET does not change the port it is listening on acording to the used protocol.
There is an environment variable that tells ROCKET on what port to listen and the default value for it is 80.
You can set it to any port you like, including 443, but that has no benefit I guess.

Relevant ROCKET variables for reference:
https://github.com/dani-garcia/vaultwarden/blob/main/.env.template#L288-L293

**Comment by LyraHarring** [comment]
You're getting a [Secure Connection Failed](https://cheapsslweb.com/blog/secure-connection-failed-firefox/) PR_CONNECT_RESET_ERROR when trying to access Vaultwarden at https://10.0.0.X:44300. This usually means the server is rejecting or abruptly closing the connection. Ensure the Vaultwarden container is serving HTTPS on port 44300 and is properly using the .crt and .key files. Also, make sure port 44300 is correctly mapped and accessible from your network. The container doesn't need /etc/ssl/certs on the host unless it relies on system CA certs. Double-check your container’s logs. Vaultwarden may not start correctly without valid certs, causing an instant reset. Hope it helps!
